[AppBar] Fix Collapse infinite feedback loop

wlemahieu opened the pull request:
Try it here: https://stackblitz.com/edit/igxjsjbv?file=src%2FDemo.tsx

Fixes #46589

# Summary

Fixes feedback loop issue when using useScrollTrigger with Collapse component inside a sticky AppBar.

The bug caused an endless expand/contract loop when scrolling, as the Collapse animation would trigger scroll position changes, which would trigger the hook again, creating a
feedback loop.

# Solution

Adds optional reentrant protection to useScrollTrigger with two new props:
- enableReentrantLock (default: false) - When true, prevents rapid state oscillations
- reentrantLockDuration (default: 300ms) - Lock duration after state changes

The fix is backwards compatible - existing behavior is preserved by default.

# Usage

For users experiencing the sticky AppBar/Collapse issue:
const trigger = useScrollTrigger({ enableReentrantLock: true });

# Testing

- ✅ All existing tests pass (14 tests)
- ✅ Added 2 new tests for reentrant protection
- ✅ Manually tested with reproduction case

Fixes [#46589](https://github.com/mui/material-ui/issues/46589)

https://github.com/user-attachments/assets/cd263f55-2f40-4f01-8f2b-0ab67b51f5f3

---

Before: https://codesandbox.io/p/devbox/angry-ramanujan-gyn43g
After: https://codesandbox.io/p/devbox/silly-leaf-9zpzrx (Using `useScrollTrigger({ enableReentrantLock: true })`)

wlemahieu committed:
fix(useScrollTrigger): add reentrant protection to prevent feedback loops with sticky AppBar and Collapse

wlemahieu committed:
prettier formatting

wlemahieu committed:
Merge branch 'master' into fix/useScrollTrigger-sticky-collapse-feedback-loop

wlemahieu committed:
drop some superfluos comments

wlemahieu committed:
Merge branch 'master' into fix/useScrollTrigger-sticky-collapse-feedback-loop

ZeeshanTamboli committed:
Merge branch 'master' into fix/useScrollTrigger-sticky-collapse-feedback-loop

ZeeshanTamboli commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.reentrant.test.js:
```diff
@@ -0,0 +1,148 @@
+    }
+  });
+
+  afterEach(() => {
+    window.pageYOffset = 0;
+  });
```
Do you need this? If yes, why?

ZeeshanTamboli commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.reentrant.test.js:
We could add these tests in the existing `useScrollTrigger.test.js` file and not create this new test file.

ZeeshanTamboli reviewed (commented):
@wlemahieu Thanks for the fix. The solution looks promising.

We should also update the docs at https://mui.com/material-ui/react-app-bar/#usescrolltrigger-options-trigger.

DenisCor commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.js:
```diff
@@ -22,26 +22,66 @@ function defaultTrigger(store, options) {
+  const {
+    getTrigger = defaultTrigger,
+    target = defaultTarget,
+    disableReentrant = true,
```
You introduced `disableReentrant = true` but your code looks only when `!disableReentrant`, consider renaming it

DenisCor commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.reentrant.test.js:
```diff
@@ -0,0 +1,148 @@
+
+      // Simulate rapid scroll events that would normally cause oscillation
+      act(() => {
+        window.pageYOffset = 101;
```
consider adding a tiny helper to avoid repetition. Example:

```
const setScroll = (y: number) =>
  act(() => {
    window.pageYOffset = y;
    window.dispatchEvent(new window.Event('scroll', {}));
  });
```

DenisCor commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.reentrant.test.js:
```diff
@@ -0,0 +1,148 @@
+import { act, createRenderer, screen, waitFor } from '@mui/internal-test-utils';
+import useScrollTrigger from '@mui/material/useScrollTrigger';
+
+describe('useScrollTrigger - reentrant protection', () => {
```
Consider also adding an unmount/cleanup test

wlemahieu committed:
update declaration file

wlemahieu committed:
cleanup, consolidation of tests

wlemahieu committed:
timeout clearing unmount test

wlemahieu committed:
Merge branch 'master' into fix/useScrollTrigger-sticky-collapse-feedback-loop

wlemahieu commented on packages/mui-material/src/useScrollTrigger/useScrollTrigger.reentrant.test.js:
```diff
@@ -0,0 +1,148 @@
+
+        // Track all trigger values
+        React.useEffect(() => {
+          triggerValues.push(trigger);
```
Good catch, thanks. This is remnant code from me testing something. Removed!

wlemahieu committed:
update app-bar documentation

wlemahieu commented:
Thank you for the excellent feedback and thorough reviews @ZeeshanTamboli & @DenisCor .

I believe I've addressed everything, but please let me know if you see anything else. Thanks.

- Updated `disableReentrant` default true to be `enableReentrantLock` default false.
- Consolidated all reentrant lock tests into main `useScrollTrigger.test.js`
- Added component unmount test to verify proper timeout cleanup, preventing memory leaks
- Implemented `setScroll` helper
- Dropped superfluos test
- Updated `app-bar.md` documentation

ZeeshanTamboli commented:
I'll assign @siriwatknp to review this further. I am a bit hesitant to add this just for an edge case.

mj12albert commented on docs/data/material/components/app-bar/app-bar.md:
```diff
@@ -126,6 +126,8 @@ A floating action button appears on scroll to make it easy to get back to the to
    - `options.disableHysteresis` (_bool_ [optional]): Defaults to `false`. Disable the hysteresis. Ignore the scroll direction when determining the `trigger` value.
    - `options.target` (_Node_ [optional]): Defaults to `window`.
    - `options.threshold` (_number_ [optional]): Defaults to `100`. Change the `trigger` value when the vertical scroll strictly crosses this threshold (exclusive).
+   - `options.enableReentrantLock` (_bool_ [optional]): Defaults to `false`. Enable a reentrant lock that prevents rapid state changes during animations. Useful when using with components like `Collapse` in a sticky `AppBar` to prevent feedback loops.
+   - `options.reentrantLockDuration` (_number_ [optional]): Defaults to `300`. Duration in milliseconds for the reentrant lock. This should match your `Collapse` component's transition timing. Only applies when `enableReentrantLock` is `true`.
```
IMO this is a rare case that should be fixed internally without creating/involve any user-facing API
